File: packages/faro-android-sdk/src/main/kotlin/com/grafana/faro/transport/TransportBody.kt
package com.grafana.faro.transport

import com.grafana.faro.api.models.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class TransportBody(
    val meta: Meta,
    val logs: List<LogEvent>? = null,
    val exceptions: List<ExceptionEvent>? = null,
    val measurements: List<MeasurementEvent>? = null,
    val events: List<EventEvent>? = null
) {
    companion object {
        private val json = Json {
            ignoreUnknownKeys = true
            encodeDefaults = false
            explicitNulls = false
        }

        fun toJson(body: TransportBody): String = json.encodeToString(body)

        fun fromJson(jsonString: String): TransportBody = json.decodeFromString(jsonString)
    }
}

enum class TransportItemType(val value: String) {
    LOG("log"),
    EXCEPTION("exception"),
    MEASUREMENT("measurement"),
    EVENT("event");
}

data class TransportItem(
    val type: TransportItemType,
    val payload: Any,
    val meta: Meta
)

File: packages/faro-ios-sdk/Sources/FaroSDK/Transport/TransportBody.swift
import Foundation

public struct TransportBody: Codable {
    public let meta: Meta
    public let logs: [LogEvent]?
    public let exceptions: [ExceptionEvent]?
    public let measurements: [MeasurementEvent]?
    public let events: [EventEvent]?

    public init(
        meta: Meta,
        logs: [LogEvent]? = nil,
        exceptions: [ExceptionEvent]? = nil,
        measurements: [MeasurementEvent]? = nil,
        events: [EventEvent]? = nil
    ) {
        self.meta = meta
        self.logs = logs
        self.exceptions = exceptions
        self.measurements = measurements
        self.events = events
    }

    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        return encoder
    }()

    private static let decoder: JSONDecoder = {
        return JSONDecoder()
    }()

    public func toJSON() throws -> Data {
        return try Self.encoder.encode(self)
    }

    public static func fromJSON(_ data: Data) throws -> TransportBody {
        return try decoder.decode(TransportBody.self, from: data)
    }
}

public enum TransportItemType: String {
    case log
    case exception
    case measurement
    case event
}

public struct TransportItem {
    public let type: TransportItemType
    public let payload: Any
    public let meta: Meta

    public init(type: TransportItemType, payload: Any, meta: Meta) {
        self.type = type
        self.payload = payload
        self.meta = meta
    }
}
